### ISSUE TITLE
[question] Device fingerprint change

### ISSUE THREAD

**Issue by krutkowski86** [oblador/react-native-keychain#336]
How do I determine that user has changed (add/remove) fingerprint/touch id/face id?

**Comment by OleksandrKucherenko** [comment]
why do you need this? Users cannot change biometric without passing the existing biometric verification first... so from security point of view, it does not matter, a user is still verified and valid.

P.S. depends on implementation on the device. Samsung just returns unique IDs for each fingerprint, but Huawei for example just returns the order number instead of unique IDs. (or vise versa, did not remember who is doing what correctly). But it's possible to detect.

**Comment by krutkowski86** [comment]
@OleksandrKucherenko 
What I meant was that someone could somehow add new fingerprint to our device which means that he can now authenticate in the app. 
Most of bank applications inform us about such change and requires standard (non biometric) authentication.

**Comment by koyta** [comment]
I’m not sure if it works but we have “Biometry current set” option. Try to find it in the docs.

**Comment by krutkowski86** [comment]
I'm using Keychain.ACCESS_CONTROL.BIOMETRY_CURRENT_SET - but it doesn't stop me from getting credentials after changing biometry settings (new fingerprint).

**Comment by krutkowski86** [comment]
@OleksandrKucherenko android docs says something about setInvalidatedByBiometricEnrollment

> // Invalidate the keys if the user has registered a new biometric
>  // credential, such as a new fingerprint. Can call this method only
>  // on Android 7.0 (API level 24) or higher. The variable
>  // "invalidatedByBiometricEnrollment" is true by default.

https://developer.android.com/training/sign-in/biometric-auth#biometric-only

Can we set it somehow?

**Comment by thijs-qv** [comment]
I don't think this is supported at the moment. Looking at the native code a new key won't be created if the old one has been invalidated.

**Comment by joelnewton** [comment]
looks like it is set to true by default? I too would like this feature.

> @OleksandrKucherenko android docs says something about **setInvalidatedByBiometricEnrollment**
> 
> > // Invalidate the keys if the user has registered a new biometric
> > // credential, such as a new fingerprint. Can call this method only
> > // on Android 7.0 (API level 24) or higher. The variable
> > // "invalidatedByBiometricEnrollment" is true by default.
> 
> https://developer.android.com/training/sign-in/biometric-auth#biometric-only
> 
> Can we set it somehow?

looks like it is set to true by default. I too would like this feature.

**Comment by jambanagar** [comment]
@krutkowski86 @joelnewton 

I set the following in CipherStorageBase.java > In the tryGenerateRegularSecurityKey() and tryGenerateStrongBoxSecurityKey() methods,

`.setUserAuthenticationRequired(true)
.setInvalidatedByBiometricEnrollment(true)
.build();`

Can you please let me know if it works for you? (assuming you are willing to fork and make changes?)

**Comment by warrioru** [comment]
any update son this topic? I use setInvalidatedByBiometricEnrollment(true) and I can still log in after adding / deleting a fingerprint.

**Comment by tieorange** [comment]
Bump. Having the same issue

**Comment by husainkantawalaFS** [comment]
any updates ?

**Comment by nguyenvantruong17** [comment]
> I’m not sure if it works but we have “Biometry current set” option. Try to find it in the docs.

It only works on iOS. But how do you know that the result you get is wrong "getGenericPassword" or "biometry" has changed because the value of "getGenericPassword" always returns "false" for both cases.

**Comment by armibit** [comment]
Any news? i've tried to add storage: Keychain.STORAGE_TYPE.RSA but I can still log in after adding / deleting a fingerprint.

**Comment by MosCD3** [comment]
same here

**Comment by mpatafio** [comment]
The current implementation doesn't support the usage of such feature for mainly one reason: the authentication is "time-bound" (setUserAuthenticationValidityDurationSeconds is set to 5 seconds), it means the Cipher can be used "asynchronously" between user authentication and encryption/decryption operations. If you want to `setInvalidatedByBiometricEnrollment` properly you have to: 
As per [documentation](https://developer.android.com/reference/android/security/keystore/KeyGenParameterSpec.Builder#setInvalidatedByBiometricEnrollment(boolean))
- `setUserAuthenticationRequired` as true AND `setUserAuthenticationValidityDurationSeconds` as **no positive value**

It basically will produce a key that requires user authentication  _every time_ you need to perform an encryption/decryption operation, it means that every time the user authenticates (trough biometry for instance) you have to perform cryptography tasks using the `Cipher` in a "synch" way (it won't be valid for a while, e.g. 5 seconds as the current implementation). 

Using the `Cipher` in a "synch" way means you have to pass it into a `CryptoObject` when you authenticate the user. 

```
protected BiometricPrompt authenticateWithPrompt(@NonNull final FragmentActivity activity) {
    final BiometricPrompt prompt = new BiometricPrompt(activity, executor, this);
    try {
      this.storage.getCachedInstance().init(Cipher.DECRYPT_MODE, context.key);
      prompt.authenticate(this.promptInfo, new BiometricPrompt.CryptoObject(this.storage.getCachedInstance()));
    }
    catch (final Throwable fail) {
      // any other exception treated as a failure
      this.onDecrypt(null, fail);
    }
    return prompt;
  }
```

Then the same `CryptoObject` will be returned as result by the `onAuthenticationSucceeded` callback.

```
 @Override
  public void onAuthenticationSucceeded(@NonNull final BiometricPrompt.AuthenticationResult result) {
    result.getCryptoObject().getCipher()
    .......
  }
```
Now the `Cipher` allows you to perform decryption of data. 

Here comes the further limitation of the current implementation: as of now `username` and `password` are encrypted/decrypted, using the scheme above your `Cipher` is unlocked only for one single operation per user authentication. Probably if we want to switch to this mode we should accept to do not encrypt `username`.

**Comment by ngominhtrungit** [comment]
I have detected fingerprint change, but still not able to faceId.

I make fingerprint change ==> app detect fingerprint change => deactived

Change the face ==> still login to the app. (result should be no login or stop it)

Has anyone done it and can guide me.

**Comment by anhquan291** [comment]
Hi everyone, Since our project requires that whenever the Biometric setting changes (like add/remove a new fingerprint) the biometric config in the app should be removed. 
I see that the lib `react-native-biometrics` can do that. But I think it would be nice if I could stick to this amazing lib. 
Any suggestions guys? Thanks

**Comment by hraschan** [comment]
Are there any updates on this?

**Comment by Daohai122** [comment]
> Hi everyone, Since our project requires that whenever the Biometric setting changes (like add/remove a new fingerprint) the biometric config in the app should be removed. I see that the lib `react-native-biometrics` can do that. But I think it would be nice if I could stick to this amazing lib. Any suggestions guys? Thanks

i am using react-native-biometrics but also not detect add a new fingerprint

**Comment by gabk17** [comment]
react-native-biometrics does not do it currently, how can that be fixed ?

**Comment by satheeshwaran** [comment]
I was trying to get this working on Android with this plugin but could not. [react-native-sensitive-info](https://mcodex.dev/react-native-sensitive-info/docs/android_options#setinvalidatedbybiometricenrollment) does this out of the box.
